### DISCUSSION TITLE
Shifting problem - coregistration Landsat BOA images

### DISCUSSION THREAD

**Discussion by albaviana** [davidfrantz/force#305] [Q&A] [answered]
Hi all,

After calculating the Landsat BAP composites for Europe I found some errors in random years and locations (England, north of the Tatras, south of Italy...) related to a "displacement" of about 5 lines of pixels northwards (I attach an example for 1998 south of London where it can be seen taking a water body as a reference). 

The strange thing is that checking the metadata of the images *MTL there is nothing weird and they were processed without problem to levels 2 and 3. I have checked those images in EarthExplorer and, although the quality is around values of 7 and 8, in some scenes you can detect a row of no-data in the middle... 
Do you have any idea how to deal with this? I would like to somehow filter those images to get rid of them at LEVEL 3 but I have not been able to find a specific pattern in the metadata or in the processing log files...

Thanks you so much in advance!!

Best

Alba.

![error_Landsat5_displacement](https://github.com/davidfrantz/force/assets/84127033/6fdf49bd-bdc7-4e43-8ebf-217f32956ff9)

**Comment by davidfrantz** [comment]
hmm,

have you tried whether the correlation score helps?

Cheers,
David

**Comment by albaviana** [comment]
Hi David,

Many thanks and sorry for the delay in replying!

After analysing each step I have not seen any problems in the processing. I have rechecked the original .tar images of those random areas and years that have given problems and the issue resides in the original images. In some of them there is a line of missing pixels or some other artefacts but it does not make a big change at image level correlations and I have not been able to detect them even with correlation score. 

They are not even 1% of > 100,000 images but if they are the ones selected in the composite at Level 3 the error is spreading. I'm still working on filtering them somehow but the metadata info didn't help either.

Best,

Alba.

**Comment by davidfrantz** [comment]
Hi @albaviana, do you still need input on this question?

**Reply by albaviana** [reply]
Hi David,

I'm working still on the issue but I don't think is anything to do with FORCE workflow but just so much noise in some images before 2000s.

Thanks for the inputs ;)

**Reply by davidfrantz** [reply]
alright...

The only thing I can think of to identify those images would be to use some sort of image matching code.

If you want to stick to FORCE, one option might be to process one single Sentinel-2 image repeatedly and each time use the different (already processed) Landsat images as base image... The logfile should ideally indicate how big the pixel shifts would be, and you could eliminate these images afterward (or probably even correct). However, this certainly is quite some workaround...

Another option would be to adapt FORCE to enable coregistration for Landsat, too... I have this on my todo list since a long time because of T2 images, but never actually approached it yet. I can't promise to deliver this soon, but I will investigate how much effort such a modification would be - with some luck, only minor changes would be necessary..

**Comment by davidfrantz** [comment, marked as answer]
Hi Alba,

it turned out to be just a few lines that needed modification.

If you want to give it a try, use the latest develop version:
https://github.com/davidfrantz/force/commit/f1272b2c391d215e3a24789d9f1fff282a476cbc

The co-registration is now also available for Landsat. You still need a base image, however.

I just ran a quick test with a tier 2 image somewhere in Germany and just used our regular base image (average NIR in 2013-2020, but renamed to 1984-2020). Results look quite promising

```
force-l2ps ~/temp/coreg-landsat/LT05_L1GS_193023_19850509_20200918_02_T2 ~/temp/coreg-landsat/germany-operational-landsat-oli.prm 
dc:  61.40%. wc:   1.13%. sc:  22.96%. cc:  65.34%. coreg (#tie, x/y shift, rmse): 253/78.50/28.28/1.16 - good, AOD: 0.2012. # of targets: 440/1744.  1 product(s) written. Success! Processing time: 03 mins 53 secs
```

Without coreg:
![image](https://github.com/davidfrantz/force/assets/32633788/4832c495-b0e6-4ebc-9b7c-525ac014516d)

With coreg:
![image](https://github.com/davidfrantz/force/assets/32633788/0c0da963-6bff-4d71-8c2a-9b734d799be9)

Cheers,
David

**Reply by albaviana** [reply]
Hi David!

Wow that's great!! Thanks for taking the time to implement it!

This will definitely solve the mismatch found. Only that at the point I'm at now, I probably won't be able to reprocess level 2 (due to limited storage capacity I had to get rid of the raw images from level 1). So, I will try the workaround for a temporary solution these days

Many thanks!

Best,

Alba
